### PR TITLE
feat: Introduce the functionality to override token_uri in credentials

### PR DESCRIPTION
**Pull request by sai-sunder-s**
There are scenarios where a custom token endpoint needs to be used. Currently the only way to provide a custom token endpoint is if a credential is created directly from the constructor. For example if a service account is loaded from a file, then custom token endpoint cannot be provided. This change introduces a method that overrides token endpoint after credential creation.

### CODE REVIEW

**Review by clundin25 — APPROVED** [review]

File: google/oauth2/credentials.py [hunk 1]
```diff
@@ -254,6 +254,23 @@ def with_quota_project(self, quota_project_id):
             enable_reauth_refresh=self._enable_reauth_refresh,
         )
 
+    @_helpers.copy_docstring(credentials.CredentialsWithTokenUri)
```
Comment (TimurSadykov) [inline google/oauth2/credentials.py:257]: Why this is needed? What credentials types it affects?

**Comment by TimurSadykov** [comment]
Please add description for the change

File: google/oauth2/credentials.py (hunk 1 above)
Comment (sai-sunder-s) [reply google/oauth2/credentials.py:257]: Python library uses helper methods to override things like quota project, scopes etc. checkout `with_quota_project` method in many credentials.
This override method is added to all the credentials that have a token uri.

Comment (TimurSadykov) [reply google/oauth2/credentials.py:257]: That is the question - do we need it for every credential type?

Comment (TimurSadykov) [reply google/oauth2/credentials.py:257]: Why the auth credetials has it as not implemented exception and here we have a default implementation? It is preferable to have explicit implementation

Comment (sai-sunder-s) [reply google/oauth2/credentials.py:257]: I am following the existing pattern https://github.com/googleapis/google-auth-library-python/blob/main/google/auth/credentials.py#:~:text=class%20CredentialsWithQuotaProject(,quota%20project.%22)

Comment (sai-sunder-s) [reply google/oauth2/credentials.py:257]: credentials.py is user credential. It is not a base class for other credentials. Base class is the credentials.py in auth package.

Comment (TimurSadykov) [reply google/oauth2/credentials.py:257]: figured out that this is in fact user creds

**Review by TimurSadykov — APPROVED** [review]
LGTM

**Pull request merged**